File: python/accel_adxl345/accel_adxl345.py
"""
accel_adxl345.py

This modules defines the AccelADXL345 class for streaming data from the
ADXL345 accelerometers. 
"""
import time
import serial
import sys
import numpy
import struct

BUF_EMPTY_NUM = 5
BUF_EMPTY_DT = 0.05

class AccelADXL345(serial.Serial):

    def __init__(self, **kwarg):

        # Command ids
        self.cmd_id = {
                'stop_streaming'       : 0,
                'start_streaming'      : 1,
                'set_timer_period'     : 2, 
                'get_timer_period'     : 3,
                'set_range'            : 4,
                'get_range'            : 5,
                'get_sample'           : 6,
                'get_max_timer_period' : 7,
                'get_min_timer_period' : 8,
                'get_bad_sample_count' : 9,
                }

        # Allowed accelerations ranges and scale factors
        self.allowedAccelRange = (2, 4, 8, 16)
        self.accelScale = 0.0384431560448

        try:
            self.reset_sleep = kwarg.pop('reset_sleep')
        except KeyError:
            self.reset_sleep = True

        try:
            self.accelRange = kwarg.pop('range')
        except KeyError:
            self.accelRange = 16
        if not self.checkAccelRange(self.accelRange):
            raise ValueError, 'unknown acceleration range {0}'.format(self.accelRange)

        _kwarg = {
                'port'     : '/dev/ttyUSB0',
                'timeout'  : 0.1,
                'baudrate' : 38400,
                }

        _kwarg.update(kwarg)
        super(AccelADXL345,self).__init__(**_kwarg)
        if self.reset_sleep:
            time.sleep(2.0)
        self.emptyBuffer()

        # Get sample dt and current range setting
        self.sampleDt = self.getSampleDt()
        self.accelRange = self.getRange()

        # Get max and min allowed sample dt
        self.minSampleDt = self.getMinSampleDt()
        self.maxSampleDt = self.getMaxSampleDt()

    def sendCmd(self,cmd):
        """
        Send the command, cmd, to the device
        """
        self.write(cmd)

    def readValue(self):
        """
        Read a value from the device.
        """
        line = self.readline()
        line = line.strip()
        return line

    def readFloat(self):
        """
        Read a single float of list of floats separated by commas
        """
        value = self.readValue()
        if ' ' in value:
            value = value.split(' ')
            value = [float(x) for x in value]
        else:
            value = float(value)
        return value

    def readInt(self):
        """
        Read a single integer or list of integers separated by commas.
        """
        value = self.readValue()
        if ' ' in value:
            value = value.split(' ')
            value = [int(x) for x in value]
        else:
            value = int(value)
        return value

    def emptyBuffer(self):
        """
        Empty the serial input buffer.
        """
        for i in range(0,BUF_EMPTY_NUM):
            #print 'empty %d'%(i,), self.inWaiting()
            self.flushInput()
            time.sleep(BUF_EMPTY_DT)


    def checkAccelRange(self,value):
        """
        Check if the value is within the allowed range set.
        """
        return value in self.allowedAccelRange

    def startStreaming(self):
        """
        Start data streaming form the accelerometer
        """
        cmd = '[{0}]\n'.format(self.cmd_id['start_streaming'])
        self.sendCmd(cmd)

    def stopStreaming(self):
        """
        Stop data streaming from the accelerometer
        """
        cmd = '[{0}]\n'.format(self.cmd_id['stop_streaming'])
        self.sendCmd(cmd)

    def getSampleDt(self):
        """
        Returns the sample interval, dt, in microseconds
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_timer_period']) 
        self.sendCmd(cmd)
        dt = self.readFloat()
        return dt 

    def getBadSampleCount(self):
        """
        Returns the number of bad/corrupted samples.
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_bad_sample_count'])
        self.sendCmd(cmd)
        val = self.readInt()
        return val

    def setSampleDt(self,dt):
        """
        Sets the sample interval in microseconds.
        """
        _dt = int(dt)
        if _dt > self.maxSampleDt or _dt < self.minSampleDt:
            raise ValueError, 'sample dt out of range'
        cmd = '[{0},{1}]\n'.format(self.cmd_id['set_timer_period'],_dt)
        self.sendCmd(cmd)
        self.sampleDt = _dt

    def getSampleRate(self):
        """
        Returns the sample rate in Hz
        """
        return 1.0/self.sampleDt


    def setSampleRate(self,freq):
        """
        Sets the sample rate in Hz
        """
        dt = int(1.0e6/freq)
        self.setSampleDt(dt)

    def getMaxSampleDt(self):
        """
        Gets the maximun allowed sample dt in microseconds.
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_max_timer_period'])
        self.sendCmd(cmd)
        value =  self.readInt()
        return value

    def getMinSampleDt(self):
        """
        Gets the minimum allowed sample dt in microseconds.
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_min_timer_period'])
        self.sendCmd(cmd)
        value = self.readInt()
        return value

    def getMaxSampleRate(self):
        """
        Returns the maximum allowed sample rate in Hz
        """
        minSampleDtSec = self.minSampleDt*(1.0e-6)
        return 1.0/minSampleDtSec

    def getMinSampleRate(self):
        """
        Returns the minum allowed samples rate in Hz
        """
        maxSampleDtSec = self.maxSampleDt*(1.0e-6)
        return 1.0/maxSampleDtSec

    def getRange(self):
        """
        Returns the current accelerometer range setting.
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_range'])
        self.sendCmd(cmd)
        accelRange = self.readInt()
        return accelRange

    def setRange(self,value):
        """
        Sets the current accelerometer range.
        """
        _value = int(value)
        if _value in self.allowedAccelRange:
            cmd = '[{0}, {1}]\n'.format(self.cmd_id['set_range'],_value)
        self.sendCmd(cmd)
        _value = self.getRange()
        self.accelRange = _value

    def getAllowedAccelRange(self):
        """
        Returns all allowed range settings
        """
        return self.allowedAccelRange

    def peekValue(self):
        """
        Grabs a sinlge sample (ax,ay,az) from the accelerometer.
        """
        cmd = '[{0}]\n'.format(self.cmd_id['get_sample'])
        self.sendCmd(cmd)
        samples = self.readFloat()
        samples = [x*self.accelScale for x in samples]
        return samples
        
    def getSamples(self,N,verbose=False): 
        """
        Streams N samples from the accelerometer at the current sample rate 
        setting.
        """
        # Start streaming 
        self.emptyBuffer()
        self.startStreaming()

        # Read samples
        data = []
        while len(data) < N:
            if verbose:
                print len(data)
            newData = self.readValues()
            data.extend(newData)

        #  Stop streaming and empty buffer
        self.stopStreaming()
        self.emptyBuffer()

        # Convert to an array, truncate to number of samples requested 
        data = numpy.array(data) 
        data = self.accelScale*data[:N,:]
        
        # Use sample rate to get array of time points
        dtSec = self.sampleDt*1.0e-6
        t = dtSec*numpy.arange(data.shape[0])

        return t, data

    #def readValues(self,verbose=False):
    #    data = []
    #    if self.inWaiting() > 0:
    #        line = self.readline()
    #        line = line.strip()
    #        line = line.split(':')
    #        for vals in line:
    #            vals = vals.split(' ')
    #            try:
    #                vals = [float(x) for x in vals]
    #                if len(vals) == 3:
    #                    data.append(vals)
    #            except:
    #                if verbose:
    #                    print 'fail'
    #    return data

    def readValues(self):
        data = []
        while self.inWaiting() >= 7:
            byteVals = self.read(7)
            ax = struct.unpack('<h',byteVals[0:2])[0]
            ay = struct.unpack('<h',byteVals[2:4])[0]
            az = struct.unpack('<h',byteVals[4:6])[0]
            chk = ord(byteVals[6]) 
            if not chk == 0:
                raise IOError, 'streaming data is not in sync.'
            data.append([ax,ay,az])
        return data

File: matlab/src/AccelADXL345.m
% AccelADXL345.m
%
% Interface to ADXL345 accelerometer connected to Arduino.
%
% Will Dickson, IO Rodeo Inc.
% -------------------------------------------------------------------------
classdef AccelADXL345 < handle
    
    properties (Constant)
        % Serial communication command ids
        cmdStopStreaming = 0;
        cmdStartStreaming = 1;
        cmdSetTimerPeriod = 2;
        cmdGetTimerPeriod = 3;
        cmdSetRange = 4;
        cmdGetRange = 5;
        cmdGetSample = 6;
        cmdGetMaxTimerPeriod = 7;
        cmdGetMinTimerPeriod = 8;
        
        % Set of allowed accelerations ranges
        allowedAccelRange = [2,4,8,16];
        
        % Scale factor for converting integers values to accelerations
        accelScale = 0.0384431560448;
    end
    
    properties
        % Serial communication parameters
        ser = [];
        port = [];
        baudrate = 38400;
        databits = 8;
        stopbits = 1;
        timeout = 1.0;
        resetDelay = 2.0;
        terminator = 'LF';
    end
    
    properties (SetAccess=private)
         % Data streaming parameters
        sampleDt = [];
        maxSampleDt = 10000;
        minSampleDt = 2000;
        range = [];
    end
    
    properties (Dependent)
        isOpen;
    end
    
    methods
        
        function self = AccelADXL345(port,baudrate)
            self.ser = serial( ...
                port, ...
                'baudrate', self.baudrate, ...
                'databits', self.databits, ...
                'stopbits', self.stopbits, ...
                'timeout',  self.timeout, ...
                'terminator', self.terminator ...
                );
        end
        
        function open(self)
            % Open serial port if not already open.
            if self.isOpen == false
                fopen(self.ser);
                pause(self.resetDelay);
            end
            self.emptyBuffer();
            self.sampleDt = self.getSampleDt();
            self.maxSampleDt = self.getMaxSampleDt();
            self.minSampleDt = self.getMinSampleDt();
            self.range = self.getRange();
        end
        
        function close(self)
            % close serial port is not already closed
            if self.isOpen == true
                fclose(self.ser);
                self.sampleDt = [];
                self.range = [];
            end
        end
        
        function delete(self)
            % Object delete function. Closes and deletes serial port
            % object.
            if self.isOpen
                self.close();
            end
            delete(self.ser);
        end
        
        function isOpen = get.isOpen(self)
            % Checks is serial port is open. Returns true (1) if it is and
            % false (0) if it is not.
            status = get(self.ser,'Status');
            if strcmpi(status,'open')
                isOpen = true;
            else
                isOpen = false;
            end
        end
        
        function dt = getMaxSampleDt(self)
            % Returns the maximum allowed sample dt
            cmd = sprintf('[%d]', self.cmdGetMaxTimerPeriod);
            self.sendCmd(cmd);
            dt = self.readNum();
        end
        
        function dt = getMinSampleDt(self)
            % Returns the minimum allowed sample dt
            cmd = sprintf('[%d]', self.cmdGetMinTimerPeriod);
            self.sendCmd(cmd);
            dt = self.readNum();
        end
        
        function dt = getSampleDt(self)
            % Get current sample dt, in microseconds, for acquiring data
            % from the device.
            cmd = sprintf('[%d]', self.cmdGetTimerPeriod);
            self.sendCmd(cmd);
            dt = self.readNum();
        end
        
        function setSampleDt(self, dt)
            % Sets the sample dt, given in microseconds, for acquiring data
            % from the device.
            if dt > self.maxSampleDt
                error('sample dt is too large');
            end
            if dt < self.minSampleDt
                error('sample dt is too small');
            end
            cmd = sprintf('[%d, %d]', self.cmdSetTimerPeriod, floor(dt));
            self.sendCmd(cmd);
            self.sampleDt = self.getSampleDt();
        end
        
        function range = getRange(self)
           % Gets the current acceleration range from the device
           cmd = sprintf('[%d]', self.cmdGetRange);
           self.sendCmd(cmd);
           range = self.readNum();
        end
        
        function setRange(self, range)
           % Sets the current acceleration range
           self.checkRange(range);
           cmd = sprintf('[%d, %d]', self.cmdSetRange, range);
           self.sendCmd(cmd);
           self.range = self.getRange();
        end
        
        function accel = peek(self)
            % Grabs a single sample from the device
            cmd = sprintf('[%d]', self.cmdGetSample);
            self.sendCmd(cmd);
            out = fscanf(self.ser);
            
            % Convert values to float array
            accel = str2array(out);
            accel = accel*self.accelScale;
            
        end
        
        function setSampleRate(self, freq)
            % Set the sample rate of the device to freq, where freq is
            % given in Hz. 
            dtSec = 1.0/freq;
            dtMicroSec = round(dtSec*1.0e6);
            self.setSampleDt(dtMicroSec);
        end
        
        function freq = getSampleRate(self)
            % Returns the current sample rate in Hz.
            dtSec = self.sampleDt*1.0e-6;
            freq = 1.0/dtSec;
        end
        
        function [data, t] = getSamples(self,n)
           % Acquire n samples from the device with the sample rate specified
           % by 1/sampleDt. 
           
           % Start Streaming
           self.emptyBuffer();
           self.startStreaming()
           
           % Collect data from device
           data = zeros(n,3);
           done = false;
           cnt = 1;
           while ~done
               outData = fread(self.ser,3, 'int16');
               outSync = fread(self.ser,1, 'uint8');
               % Check that last byte is zero - if not we are out of sync
               if ~(outSync == 0) 
                   error('Error reading data - stream out of sync');
               end
               data(cnt,:) = [outData(1),outData(2),outData(3)];
               cnt = cnt + 1;
               if cnt >= n
                   done = true;
               end
           end
           
           % Stop Streaming and empty buffer
           self.stopStreaming();
           self.emptyBuffer();
           
           % Convert integer data to acceleration and create time array
           data = self.accelScale*data;
           dtSec = self.sampleDt*1.0e-6;
           t = [1:n]*dtSec;
        end
        
        %end
        % -----------------------------------------------------------------
        %methods (Access=private)
        
        function emptyBuffer(self)
            % Empty serial port buffer
            if self.isOpen
                if self.ser.BytesAvailable > 0
                    fread(self.ser, self.ser.BytesAvailable);
                end
            end
        end
        
        function sendCmd(self,cmd)
            % Send command to serial serice
            if self.isOpen
                fprintf(self.ser,'%s\n',cmd);
            else
                error('serice must be open to send message');
            end
        end
        
        function val = readNum(self)
            % Reads a single number from the serial port.
            outStr = fscanf(self.ser);
            val = str2num(outStr);
        end
        
        function stopStreaming(self)
            % Stop serice streaming
            cmd = sprintf('[%d]', self.cmdStopStreaming);
            self.sendCmd(cmd);
        end
        
        function startStreaming(self)
            % Start serice streaming
            cmd = sprintf('[%d]', self.cmdStartStreaming);
            self.sendCmd(cmd);
        end
        
        function checkRange(self,val)
           % Check whether or not the value is within the set of allowed
           % acceleration ranges.
           if ~any(self.allowedAccelRange == val)
               error('acceleration range not allowed');
           end
           
        end
        
    end
    
end


% -------------------------------------------------------------------------
function array = str2array(line)
% Converts an input string of numbers separated by commas into to an array.

line = strtrim(line);
line = strsplit(line,' ');
array = zeros(size(line));

% Convert elements to doubles
test = true;
for i = 1:length(line)
    array(i) = str2double(line{i});
    if isempty(array(i))
        test = false;
    end
end
% Return empty array if read failed
if test == false
    array = [];
end
end
